### PR TITLE
Fix shutdown race

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/IISIntegration/issues/949

### CODE REVIEW

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 1]
```diff
@@ -38,8 +40,32 @@ IN_PROCESS_HANDLER::OnExecuteRequestHandler()
             L"InProcess Application");
     }
 
-    //SetHttpSysDisconnectCallback();
-    return m_pApplication->OnExecuteRequest(m_pW3Context, this);
+    if (m_pRequestHandler == NULL)
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Good change 👍

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 2]
```diff
@@ -49,42 +75,36 @@ IN_PROCESS_HANDLER::OnAsyncCompletion(
...
-    return m_pManagedHttpContext;
+    // Call the managed handler for async completion.
+    return m_pAsyncCompletionHandler(m_pManagedHttpContext, hrCompletionStatus, cbCompletion);
 }
 
-BOOL
-IN_PROCESS_HANDLER::QueryIsManagedRequestComplete(
-    VOID
-)
+REQUEST_NOTIFICATION_STATUS IN_PROCESS_HANDLER::ReturnServerShutdown() const
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Rename this to ServerShutdownMessage().

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 3]
```diff
@@ -673,7 +586,7 @@ IN_PROCESS_APPLICATION::CreateHandler(
     HRESULT hr = S_OK;
     IREQUEST_HANDLER* pHandler = NULL;
 
-    pHandler = new IN_PROCESS_HANDLER(pHttpContext, this);
+    pHandler = new IN_PROCESS_HANDLER(CreateReferencedApplication(this), pHttpContext, m_RequestHandler, m_RequestHandlerContext, m_AsyncCompletionHandler);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: Rename `CreateReferencedApplication` to something else as you aren't creating the application.

File: src/AspNetCoreModuleV2/CommonLib/iapplication.h [hunk 4]
```diff
@@ -54,3 +55,18 @@ class IAPPLICATION
...
+struct IAPPLICATION_DELETER
+{
+    void operator ()(IAPPLICATION* application) const
+    {
+        application->DereferenceApplication();
+    }
+};
+
+template< class APPLICATION >
+std::unique_ptr<APPLICATION, IAPPLICATION_DELETER> ReferenceApplication(APPLICATION* application)
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/CommonLib/iapplication.h]: Do we need to do this for out of process? Also doesn't this more logically belong as part of `Application`?

Comment (pakrym) [reply src/AspNetCoreModuleV2/CommonLib/iapplication.h]: We need it for shim

**Pull request merged**